### PR TITLE
Add loss_type='weighted_nll' for per-sample loss scaling in SFT

### LINKED ISSUE #5761: Weighted SFT
### Method description

I was in the process of writing a weighted training setup for SFT and thought it may add value to the TRL package. 

Normal SFT treats all assistant tokens equally. I personally found this frustrating because it makes mixed quality long form SFT less than ideal in a lot of ways. There is no real middle ground between a purely RL focused method or SFT focused method. The following implementation handles this at the assistant response level, but in the future I think adding the token level dynamics are perfectly feasible, and it's more an issue of data formatting. 

I'm proposing to add an OPTIONAL weight value to the JSON. It's a scalar -1 to 1 value. I expect the typical negative range not to go below -0.2 as negative values should be intentionally small. 

Something like the following :

{
  "messages": [
    {"role": "user", "content": "..."},
    {"role": "assistant", "content": "...", "weight": 1.0},
    {"role": "user", "content": "..."},
    {"role": "assistant", "content": "...", "weight": -0.1} 
  ]
}

The weight acts as a loss scalar. Positive values reinforce the response proportionally, while negative values invert the gradient direction. The above example of -0.1 acts to gently push the model away from that completion. 

It should be entirely backward compatible as this codepath only activates if a weight field is present in the dataset. I can separate the codepath such that the normal loss function logic is entirely untouched. Whatever is preferred. 

In terms of memory and compute versus standard SFT, there's pretty negligible overhead. 

A 20 turn roleplay conversation where turns 1 through 18 are mediocre but turns 19 and 20 contain the behavior you actually want to reinforce can now be used without contaminating your gradient with the mediocre turns.

Testing plan :

Unit test: verify `loss_weights` tensor is correctly produced by the collator from message-level `weight` fields
Unit test: verify weighted loss scales proportionally with weight magnitude
Integration test: train a small model with negative weights on a target token and verify suppression

Reference :
Reward-weighted regression applied to per-turn granularity in multi-turn conversations. Structurally parallels GRPO/RWR in data flow while using weighted cross-entropy instead of policy gradient.

This effectively extends the concept of OpenAI's fine tuning API from binary 0/1 weight logic per assistant message in to a continuous value. 

### Open source status

- [ ] The method implementation is available
- [ ] The model weights are available
- [ ] The training datasets are available

### Provide useful links for the implementation

I plan to write the implementation (~50 lines), mostly reusing existing SFT loss logic with a per-turn scalar applied. No external paper or reference implementation exists.  The closest conceptual precedent is reward-weighted regression applied at per turn granularity. Happy to open a PR directly.

### PR DESCRIPTION
**Pull request by Sumu004**
# What does this PR do?

Adds `loss_type="weighted_nll"` to `SFTTrainer`, letting each training example carry a float `"weight"` column that scales its contribution to the loss. This is a lightweight middle ground between standard SFT (all samples equal) and a full RL loop.

Fixes #5761

---

## Motivation

Standard SFT treats every assistant token equally. In practice datasets are mixed-quality: a 20-turn conversation where turns 1–18 are mediocre but turns 19–20 contain the behaviour you want shouldn't pollute the gradient equally. With `weighted_nll` you can express this directly in the dataset:

```python
{"messages": [...], "weight":  1.0}   # full reinforcement (default)
{"messages": [...], "weight":  0.5}   # half strength
{"messages": [...], "weight": -0.1}   # gentle gradient inversion
{"messages": [...], "weight":  0.0}   # completely ignored
```

```python
trainer = SFTTrainer(
    model="...",
    args=SFTConfig(loss_type="weighted_nll", ...),
    train_dataset=dataset_with_weight_column,
)
```

**How the loss is computed:**

```
per_sample_loss_i = mean_{valid tokens} CE(logits_i, labels_i)
loss = Σ (per_sample_loss_i × weight_i) / Σ |weight_i|
```

Normalising by `Σ|w|` keeps the effective learning rate stable regardless of weight distribution. When `num_items_in_batch` is provided (gradient accumulation path), the denominator switches to the global token count to match the existing NLL behaviour.

**`weight=1.0` everywhere is numerically identical to `loss_type="nll"` — unit-tested.**

---

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. → #5761
- [x] Did you make sure to update the documentation with your changes? (docstring in `SFTConfig`)
- [x] Did you write any new necessary tests? (`TestWeightedNLLLoss` with 6 unit tests + integration test `test_train_weighted_nll_loss`)

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@qgallouedec @lewtun — this directly addresses the request in #5761.

---

> [!NOTE]
> **Medium Risk**
> Changes core SFT loss computation and training gradients; negative weights invert updates, but behavior is guarded, documented, and covered by tests.
> 
> **Overview**
> Adds **`loss_type="weighted_nll"`** to `SFTTrainer` so each example can supply a float **`weight`** column that scales its contribution to the training loss (positive reinforce, negative invert gradient, zero skip). Loss is per-sample mean CE, aggregated as **Σ(wᵢ × lossᵢ) / Σ|wᵢ|** (or token-normalized when `num_items_in_batch` is set for gradient accumulation).
> 
> **`weighted_nll_loss`** is wired in **`compute_loss`** (weights are popped before the model forward). Data collators pass **`weight`** through batches; **`weight`** is kept in signature columns when `remove_unused_columns` is on. **`SFTConfig`** documents the new option.
> 
> **Guards:** incompatible with **`packing`**, **`padding_free`**, **`use_liger_kernel`**, custom **`compute_loss_func`**, and CP/SP **`shift_labels`** paths; missing **`weight`** in the batch raises a clear error. MoE **router aux loss** is re-added after replacing the default loss.
> 
> Tests add **`TestWeightedNLLLoss`** (six unit tests) and **`test_train_weighted_nll_loss`** integration.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 349a4cb2fe5185b916fddc652e71dcfe001c76d7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -812,6 +812,66 @@ def _collate_prompt_completion(self, examples: list[dict[str, Any]]) -> dict[str
...
+    weights = weights.to(dtype=per_sample_loss.dtype, device=per_sample_loss.device)
+    weighted_loss = (per_sample_loss * weights).sum()
+
+    if num_items_in_batch is None:
+        denom = weights.abs().sum().clamp(min=1e-8)
+        loss = weighted_loss / denom
+    else:
+        if isinstance(num_items_in_batch, torch.Tensor):
+            num_items_in_batch = num_items_in_batch.to(weighted_loss.device)
+        loss = weighted_loss / num_items_in_batch
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py]: true

**Review by qgallouedec — COMMENTED** [review]
Thanks, do you have any reference work/paper where this is used?

**Comment by Sumu004** [comment]
Thanks for the review! Addressed all 5 issues from Cursor Bugbot in the latest commit:

1. **Collator drops weight** → `torch_call` now collects `weight` from each example and stacks it into `output["weight"]`
2. **Liger incompatibility** → raises `ValueError` (same pattern as `chunked_nll`)
3. **Packing incompatibility** → raises `ValueError` at init; packing merges examples so per-example weights are semantically undefined
4. **shift_labels / CP-SP** → raises `ValueError` in `compute_loss` when `labels is None`
5. **num_items_in_batch scaling** → in the grad-accumulation path, broadcasts the per-sample weight to every token and divides by global token count, matching the units of `loss_type='nll'`

---

Re: reference papers — per-sample loss weighting during SFT is used in several lines of work:

- **DoReMi** (Xie et al., 2023 — https://huggingface.co/papers/2310.03025): optimises domain mixture weights for LM pretraining; each domain gets a scalar weight updated via online DRO
- **LIMA** (Zhou et al., 2023 — https://huggingface.co/papers/2305.11206) and follow-up **data-quality weighting** papers use quality scores as sample weights to down-weight noisy demonstrations
- **Reward-weighted regression / RAFT** (Dong et al., 2023 — https://huggingface.co/papers/2304.06767): scales the SFT cross-entropy of each completion by its normalised reward, which is exactly a per-sample weight on the NLL
- **Negative samples / contrast SFT** (e.g. ORPO, SimPO): negative examples get weight < 0 so the gradient points away from them — the signed-weight convention in this PR supports that directly

The common thread is: rather than filtering or up/down-sampling the dataset, you keep every example in the batch but modulate its gradient contribution with a scalar, which is simpler and compatible with packing-free distributed training.

**Comment by RyanJFriedrich** [comment]
@Sumu004 

Thanks for moving to add this in. It slipped me in terms of working the pull request.

I had the code working in my fork of this. Similarly to what you appear to have done. 
I patched the code path in the way I did because I needed it for training.

It's an extremely nice feature to have for training. It exists in OpenAI's literature as noted in my issue but this treats it as scalar instead of binary - which I needed for some research.

I experimented with various types of offline RL via this method where the model temporarily "leaves policy" for a short duration and the policy gets regenerated offline and trained. It allows language models to cover a significant chunk of tasks modern RL won't handle in terms of depth of strategy because you can omit failed strategy and highlight working strategy as easily. It creates signal dilution. 

Very minor negative weights showed a strong ability to teach a model to speak a different way. I tested this with simple 2 or 3 turn outputs where a model gave ground truth, scored at -0.05. Then positive variants of the same turns, all set at 1 weight.

This allowed a simple study where Gemma 4 could be made "Claude like" in the way it spoke by having Claude's outputs as SFT data.

**Comment by Sumu004** [comment]
@RyanJFriedrich Thanks so much for sharing this, it's genuinely exciting to hear from someone who's already put this to use in real research. The scalar-weight framing (vs. binary masking) was exactly the motivation: binary filtering throws away the sample entirely, but a continuous weight lets you tune how much gradient each example contributes, which opens up the offline RL workflows you describe without any extra infrastructure.

The negative-weight experiments sound fascinating, using -0.05 on ground-truth outputs to steer style while keeping strong positive signal on target examples is a really elegant use of the mechanism. I hadn't thought to apply it that way but it makes complete sense: the model learns both what to say and how to say it from a single loss, rather than needing a separate style objective.

Would you be open to sharing a snippet of how you structured the dataset or the weight schedule for the offline RL case? Even informally, it'd be a great usage example to point to if the maintainers want concrete motivation for merging this.

**Comment by RyanJFriedrich** [comment]
@Sumu004 

Sure. I detailed it in the review on the 0.5b model. 

Using this style of RL is interesting because you re-frame the "policy" to become the actual dataset you generate.

I used multiple passes of Gemma 4 31b.

1. Design a hard problem. I used Zebra puzzles. 
2. Let the model attempt to solve.
3. Let the model do post hoc analysis of how it succeeded or failed.
4. Let OTHER models analyze and assign scores based on rubric.
5. Run the SFT. Set the problem to 0 weight. 

This attempts to defeat single model bias. If multiple models agree that an introspective step was correct you positively weight it. Or if it was confabulation or partial confabulation, etc. 

You're masking the problem / solution itself because you want to prevent memorization of the problem. You're focused on how the model attends to the problem.

At large scale, if a small percentage of negative weighting exists, it's okay because enough positive signal exists to smooth the geometry of the neural network. Even using a -0.01 weight is enough to gently nudge the model away from a response. This can be done with pure confabulation or outright hallucination. 

You're working on the model's meta-cognitive ability on some level in the example above. It was just one of the tested use cases that I found interesting. You can attempt to test for memorization but because the model gets -100 tokens across the solution to the problem, it genuinely doesn't have the full problem. It only has a sense of the type of problem. 

I had some code for generating these puzzles then using a SAT solver to make them maximally hard.

I wasn't comfortable writing some of the code for this repo because their submission policy. I could write their LIGER code, etc but I would basically NEED Opus to write it and I cannot confidently say "Yes I fully understand." ... 

Lastly, I'm going to probably submit some datasets to HuggingFace so that people can see how it all works with a simple research.md or something. This was for another project, but I think it's valuable because the idea is simple, and anyone can use it. Perhaps I'll poke the Unsloth people to implement it after the datasets exist and if your PR gets handled here. 

Thanks!

**Comment by qgallouedec** [comment]
Thanks @Sumu004, clean, well-tested implementation.

The gate here is scope. My request for prior work is still effectively open, demand is thin (only #5761, no other askers / reactions), and @matdou flagged that this overlaps closely with reward-weighted SFT / RWR (ASFT) and token-level Forgetting-SFT. So it'd help to position clearly why a hand-set scalar weight earns its place over those.

Two concrete things if it's accepted:
- #5761 actually asked for *per-turn* weights on individual assistant messages, but this PR weights the *whole example*, worth being explicit about that narrower scope;
- please drop the defensive `getattr(...)` calls in compute_loss (repo style avoids them) and fix `test_zero_weight_suppresses_sample`, which compares two identical weight tensors and so passes trivially.

Holding on the scope decision before a deeper review. If demand remain thin, it will likely be closed.

**Comment by Sumu004** [comment]
@qgallouedec Thanks for the detailed gate — addressing both concrete items and the scope question.

**Technical items (done, pushed):**
- Dropped all `getattr(...)` calls in the MoE aux-loss restore — `self.aux_loss_enabled` already gates on a real MoE config, so `text_config`/`router_aux_loss_coef` are guaranteed present; switched to direct attribute access and `self._is_vlm` (matching the existing `_chunked_ce_forward` mirror).
- Fixed `test_zero_weight_suppresses_sample` — it now compares the full batch (sample 0 weighted to zero) against the second sample alone (batch size 1, weight 1.0), which actually exercises suppression instead of comparing a tensor to itself.

**Scope: per-turn vs whole-example, and why this earns its place next to RWR/ASFT and Forgetting-SFT**

You're right that #5761 asked for per-turn weights and this PR is narrower — whole-example only. I'd rather ship the narrower, simpler thing now than block on the harder per-turn design (which needs a different collator contract: weights aligned to assistant-turn boundaries, not one scalar per row). Happy to scope the PR title/docs explicitly to "per-example" and treat per-turn as a documented follow-up rather than conflate them here.

On ASFT/Forgetting-SFT: both compute weights *for* you (RWR from a reward model, Forgetting-SFT from a second backward pass). This PR doesn't compute weights — it's the plumbing that lets a *precomputed* weight (from any source: a reward model, a human label, a heuristic, or nothing more than a JSONL column) reach the loss. ASFT and Forgetting-SFT could both be built on top of this primitive instead of each rolling their own per-sample scaling.

Concretely, from a maintainer of this exact feature's actual use ([RyanJFriedrich's comment above](#)):
- Used today to train Gemma 4 31B with a 5-step self-critique pipeline: model attempts a hard problem, does post-hoc analysis, multiple other models score the analysis by rubric, and the *problem itself* gets masked (weight 0) while the analysis gets a hand-set scalar — explicitly to avoid baking in single-model judge bias.
- A second documented experiment: weighting undesired completions at -0.05 against desired ones at 1.0 measurably shifted token-level preferences (style transfer) without destabilizing unrelated outputs — small negative weight, large positive base, no separate contrastive objective needed.

Neither workflow fits ASFT (offline, frozen reference weights) or Forgetting-SFT (requires training a second model to compute token-level attribution) — both need a *hand-set*, externally-computed weight per example, which is exactly what's missing today. The whole feature is ~150 lines because it's intentionally just plumbing, not a new training algorithm.

Let me know if a doc note distinguishing this from RWR/ASFT/Forgetting-SFT (and explicitly flagging per-turn as future work) would help close this out.

**Comment by RyanJFriedrich** [comment]
@Sumu004 RE: Per prompt reward signaling, it's in my code repo. You can set individual weight per prompt in my code example provided via the weighted SFT function and the way it handles data collation. 

@qgallouedec I can assist with the value difference. We sit on a massive trove of SFT data on HF. All of that data can be weighted. We're treating every token as equally valuable and I think it's a mistake. I gave my objection to the other methods in the issue. The data doesn't necessarily need to be hand weighted either. Modern models can efficiently weight data given a rubric. Multiple models help bring down per-model bias. 

Per token and other frozen state models are superior in many senses. The issue becomes the resource utilization. Weighted SFT as proposed requires very little extra to pull cleaner signal from massive existing datasets. 

If providing a dataset would help change your mind or something, I can provide one of my research datasets via HF. The way it's written in my own implementation creates zero overhead to anyone who just wants to run "normal" SFT versus "weighted". I wasn't comfortable with submitting my code for PR because it would require features I'd use an LLM to write and I was trying to follow strict guideline on "Minimal LLM" code while also being 100% able to defend any line written.

I hope that offers clarity on why I never wrote the PR during my research.

**Comment by Sumu004** [comment]
@qgallouedec following up on the scope question, since it's been about two weeks since I addressed the getattr cleanup, the trivial test fix, and the positioning against RWR/ASFT and Forgetting-SFT. Let me know if there's anything else you'd like adjusted, or if you've reached a decision on scope. Thanks!

**Comment by qgallouedec** [comment]
closing, see https://github.com/huggingface/trl/issues/5761#issuecomment-5053042966
